lib/puppet/pops/merge_strategy.rb

```diff
@@ -0,0 +1,218 @@
+    end
+
+    def checked_merge(e1, e2, options)
+      raise NotImplementedError.new('Subclass must implement checked_merge(e1,e2,options)')
+    end
+
+    def assert_type(param, type, value)
+      TypeAsserter.assert_instance_of(param, type, value)
+    end
+  end
+
+  # Performs a native Hash merge
+  #
+  class HashMergeStrategy < MergeStrategy
+    # Merges hash e1 onto hash e2 in such a way that if the values of duplicate keys
+    # will be those of e1
+    #
+    # @param e1 [Hash<String,Object>] The hash that will act as the source of the merge
+    # @param e2 [Hash<String,Object>] The hash that will act as the receiver for the merge
+    # @return [Hash<String,Object]] The merged hash
+    # @see Hash#merge
+    def checked_merge(e1, e2, options)
+      e2.merge(e1)
+    end
+
+    def key
+      :hash
+    end
+
+    protected
+
+    def value_t
+      @value_t ||= Puppet::Pops::Types::TypeParser.new.parse('Hash[String,Data]')
+    end
+
+    MergeStrategy.add_strategy(self)
+  end
+
+  # Merges two values that must be either scalar or arrays into a unique set of values.
+  #
+  class UniqueMergeStrategy < MergeStrategy
+    # Merges two values that must be either scalar or arrays into a unique set of values.
+    #
+    # Scalar values will be converted into a one element arrays and array values will be flattened
+    # prior to forming the unique set. The order of the elements is preserved with e1 being the
+    # first contributor of elements and e2 the second.
+    #
+    # @param e1 [Array<Object>] The first array
+    # @param e2 [Array<Object>] The second array
+    # @return [Array<Object>] The unique set of elements
+    #
+    def checked_merge(e1, e2, options)
+      to_flat_a(e1) | to_flat_a(e2)
+    end
+
+    def key
+      :unique
+    end
+
+    protected
+
+    def value_t
+      @value_t ||= Puppet::Pops::Types::TypeParser.new.parse('Variant[Scalar,Array[Data]]')
+    end
+
+    def to_flat_a(e)
+      e.is_a?(Array) ? e.flatten : [e]
+    end
+
+    MergeStrategy.add_strategy(self)
+  end
+
+  # Performs a deep merge. The arguments can be either Hash or Array
+  #
+  # @see https://github.com/danielsdeleo/deep_merge
+  #
+  class DeepMergeStrategy < MergeStrategy
+    def checked_merge(e1, e2, options)
+      # e1 (the destination) is cloned to avoid that the passed in object mutates
+      DeepMerge.deep_merge!(e1, e2.clone, { :preserve_unmergeables => false }.merge(options))
+    end
+
+    def key
+      :deep
+    end
+
+    protected
+
+    # Returns a type that allows all deep_merge options except 'preserve_unmergeables' since we force
+    # the setting of that option to false
+    #
+    # @return [Puppet::Pops::Types::PAnyType] the puppet type used when validating the options hash
+    def options_t
+      @options_t ||= Puppet::Pops::Types::TypeParser.new.parse('Struct[{'\
+          "strategy=>Optional[Pattern[#{key}]],"\
+          'knockout_prefix=>Optional[String],'\
```

Comment:
These options should be in the documentation - we have selected one deep merge strategy, I don't think we will have others.